### ISSUE TITLE
Update templates to support stages tag

### ISSUE THREAD

**Issue by JordanSussman** [go-vela/community#233]
## Description

Templates should support the [stages](https://go-vela.github.io/docs/concepts/pipeline/stages/) tag.

## Value

Allow users of templates to further reduce duplication between the source `.vela.yml` and template(s).

## Definition of Done

Users are able to define the `stages` block within any template.

## Impacted Personas (Optional)

Users of templates

**Comment by 10hendersonm** [comment]
I tried to build a staged matrix build template today and this was ultimately my downfall.

**Comment by JordanSussman** [comment]
Brief documentation of the pipeline type functionality was added [here](https://go-vela.github.io/docs/templates/#templating-directly-in-velayml).

**Comment by JordanSussman** [comment]
I'm thinking the design for this would look something like this:

```yaml
version: "1"

templates:
  - name: go
    source: github.com/octocat/hello-world/.vela/build.yml
    format: go
    type: github
    render_inline: true # defaults to false
    vars:
      image: golang:latest
```

**note** the addition of the `render_inline` and `vars` underneath `templates` to allow rendering a template without needing to reference it underneath `stages` or `steps` later in the file.

With the `build.yml` file looking like:

```yaml
metadata:
  template: true

stages:
  test:
    steps:
      - name: test
        image: {{ .image }}
        ruleset:
          event: [ push, pull_request ]
        commands:
          - go test ./...
  build:
    steps:
    - name: test
      image: {{ .image }}
      ruleset:
        event: [ push, pull_request ]
      commands:
        - go build
```

The ability to utilize `render_inline: true` with `stages` and `steps` outside of the `.vela.yml` would be explicitly disallowed in this design, so the following file would return an error when submitted to the server:

```yaml
version: "1"

templates:
  - name: go
    source: github.com/octocat/hello-world/.vela/build.yml
    format: go
    type: github
    render_inline: true # defaults to false
    vars:
      image: golang:latest


steps:
  - name: echo
    image: alpine
    commands:
      - echo foo
 ```

**Comment by jbrockopp** [comment]
Started a POC for this:

https://github.com/go-vela/server/tree/poc/stages/templates
https://github.com/go-vela/types/tree/poc/stages/templates

**Comment by JordanSussman** [comment]
Utilizing the `.vela.yml` within https://github.com/go-vela/template-poc/blob/main/.vela.yml with https://github.com/go-vela/server/pull/597 results in the following build:

![image](https://user-images.githubusercontent.com/8935518/155809195-a40d3dd2-fdec-4a81-936d-c672a05cb82a.png)

@kneal and I are hopefully going to get the code finished for review sometime next week.

**Comment by jbrockopp** [comment]
This functionality has been accepted into the Vela codebase:

* https://github.com/go-vela/cli/pull/341
* https://github.com/go-vela/server/pull/597
* https://github.com/go-vela/server/pull/608
* https://github.com/go-vela/types/pull/237

It will be available in the `v0.13.x` release of Vela.
